### DISCUSSION TITLE
[storer] retrieve stored object list pre commit

### DISCUSSION THREAD

**Discussion by guykatz** [microstream-one/microstream#670] [Q&A] [answered]
Hi all;
given a storer that add several objects to be commited.
after all objects in this "transaction" are stored but before they are commited, is it possible to retrieve this list of Objects?
I playing with the idea of adding dynamic proxy that will do something when storer.commit is called but I need access to the object list that are going to be commited in order to fullfill my proxy functionality.
thanks

**Comment by hg-ms** [comment, marked as answer]
Hi,

there is `one.microstream.persistence.types.PersistenceLocalObjectIdRegistry.iterateMergeableEntries(PersistenceAcceptor)` which is implemented by our default storers. As it is not part of the `one.microstream.persistence.types.Storer` interface you need to cast the storer instance to ` PersistenceLocalObjectIdRegistry` or to the storer implementation itself e.g. `one.microstream.persistence.binary.types.BinaryStorer.Default` 

```
final BinaryStorer.Default storer = (BinaryStorer.Default)storage.createLazyStorer();

storer.store(...);
storer.iterateMergeableEntries(
	(id, object) ->
	System.out.println("id: " + id + " object: " + object.getClass()));
storer.commit();
```

**Reply by guykatz** [reply]
thanks @hg-ms 
from the eclipse store architecture perspective, where is the correct place to plug a dynamic proxy of sorts that will intercept the commit method?
is doable with foundation configuration when creating the elcipse store objects? or do I need to change some eclipse store code?
looking at the API the only way that might do the trick is through embbededstoragefoundation decorate PersistenceLiveStorerRegistry.registerStorer​(PersistenceStorer storer)
and also decorate PersistenceStorer  when registering (so 2 dynamic proxies or something like that)
does this make sense?

**Reply by hg-ms** [reply]
I think the best way would be to create custom storers (Lazy and Eager) by extending the existing Default storer (see BinaryStorer.java).
They can be set up by a custom ` PersistenceStorer.Creator` that can be configured by the connectionFoundation:
```
final EmbeddedStorageManager storage = EmbeddedStorage.Foundation(WORKDIR)
	.onConnectionFoundation(cf -> cf.setStorerCreator(new CustomStorerCreator(
			cf.getStorageSystem().channelCountProvider(),
			cf.isByteOrderMismatch()
		)))
	.start();
```
